maven/poms.bzl

```diff
@@ -45,38 +65,223 @@ def _process_dependency(dep_node):
+
+# Description:
+#   Merges the dependency section of the pom.  This makes an assumption that deps sections won't have both the main
+#   artifact for a group_id/artifact_id pair AND one of the same pom's classified artifacts.  It is possible, and in
+#   those cases, the deps will be wrong and the build snippet should be explicitly substituted.
+def _merge_dependency_section(parent, child):
+    if not bool(parent):
+        return child if bool(child) else xml.new_node(label = labels.DEPENDENCIES)
+    elif not bool(child):
+        return parent if bool(parent) else xml.new_node(label = labels.DEPENDENCIES)
+    if parent.label != labels.DEPENDENCIES:
+        fail("Parent node in merged dependency operation not a <dependencies> node: %s" % parent)
+    elif child.label != labels.DEPENDENCIES:
+        fail("Child node in merged dependency operation not a <dependencies> node: %s" % child)
+
+    # index the <dependency> nodes by groupId:artifactId
+    parent_deps = {}
+    for node in _children_if_exists(parent):
+        key = "%s:%s" % (xml.find_first(node, labels.GROUP_ID), xml.find_first(node, labels.ARTIFACT_ID))
+        parent_deps[key] = node
+    child_deps = {}
+    for node in _children_if_exists(child):
+        key = "%s:%s" % (xml.find_first(node, labels.GROUP_ID), xml.find_first(node, labels.ARTIFACT_ID))
+        child_deps[key] = node
+
+    merged = {}
+    for key in parent_deps:
+        merged[key] = parent_deps[key]
+    for key in child_deps:
+        if bool(merged.get(key, None)):
+            existing_node = merged[key]
+            merged[key] = xml.new_node(
+                label = labels.DEPENDENCY,
+                children = _merge_leaf_elements(
+                    _children_if_exists(merged[key]),
+                    _children_if_exists(child_deps[key]),
+                ),
+            )
+        else:
+            merged[key] = child_deps[key]
+
+    dependency_items = []
+    for key, node in merged.items():
+        dependency_items.append(node)
+    return xml.new_node(
+        label = labels.DEPENDENCIES,
+        children = dependency_items,
+    )
+
+# A highly constrained merge (only merges core metadata, properties, dependency_management, and dependency sections.
+# It drops all other sections on the floor, including parent pom metadata.  It is also not as efficient as it could be,
+# because pom sections are unordered, so there's a lot of scanning.  It also requires lots of targetted methods since
+# starlark has no recursion, so this code cannot be generalized without becoming a hellish batch of insane iteration.
+def _merge_parent(parent, child):
+    merged = xml.new_node(label = labels.PROJECT, children = [])
+
+    # merge core identity metadata
+    for label in [labels.GROUP_ID, labels.ARTIFACT_ID, labels.VERSION]:
+        merged_node = _merge_content_last_wins(xml.find_first(parent, label), xml.find_first(child, label))
```

Comment:
I assume this policy (last wins) is consistent with out maven would resolve such ambiguity?

Reply:
Yeah.  Between parent->child the child's metadata wins, for a given leaf element.  It's not ambiguous, it's precisely the strategy for metadata merging in poms.